### PR TITLE
Wrapper for Varembed Models

### PR DESCRIPTION
**Pull request by anmol01gulati**
This is an effort to integrate [Varembed word-embedding Model](https://arxiv.org/abs/1608.01056) into gensim. 
To train varembed models, one can use the code put up by @rguthrie3 [here](https://github.com/rguthrie3/MorphologicalPriorsForWordEmbeddings).
Presently, the motive is to provide wrapper for loading varembed models with gensim and not training the word embeddings.

The first draft is based on Wordrank wrapper by @parulsethi 

TODOs
- [x] Add tests for varembed wrapper.
- [x] Add model files for testing loading of wrapper.
- [x] Add tutorial in documentation.
- [x] Add RST files for Sphinx APIREF autogeneration.

### CODE REVIEW

**Comment by anmol01gulati** [comment]
I've added wrapper to load varembed model into gensim and allowing keyedvectors functionalities. 
As the wrapper includes dependency over morfessor package. I've factored that out into a different method and ensemble the morpheme embeddings iff the user asks for ensembling the morpheme embeddings and morfessor package is present on the system.
@tmylk @piskvorky Please review.

**Comment by anmol01gulati** [comment]
There seems to be a small issue specifically in the morfessor package. I've submitted a [PR](https://github.com/aalto-speech/morfessor/pull/6) for the same. We'll probably have to wait to get it merged or find some other fix to get the code running in Python 2.6

**Comment by anmol01gulati** [comment]
@tmylk Apart from the issue Python 2.6, which I've already discussed. The code right now fails in other versions as well, which seems to me a problem in circular dependency in imports in morfessor. Though the tests run fine on my personal machine, it fails on Travis, which is something I don't understand. Any ideas, what's going wrong?

File: gensim/models/wrappers/varembed.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        self.wv = KeyedVectors()
+        self.vector_size = 0
+        self.vocab_size = 0
+
+    @classmethod
+    def load_varembed_format(cls, vectors, morfessor_model=None, use_morphemes=False):
+        """
+        Load the input-hidden weight matrix from the fast text output files.
+
+        Note that due to limitations in the FastText API, you cannot continue training
```
Comment (tmylk) [inline gensim/models/wrappers/varembed.py]: Please correct Docstring to be about varembed

File: gensim/models/wrappers/varembed.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+from gensim.models.word2vec import Word2Vec
+
+# utility fnc for pickling, common scipy operations etc
+from gensim import utils
+from gensim.models.word2vec import Vocab
+
+logger = logging.getLogger(__name__)
+
+
+class VarEmbed(Word2Vec):
```
Comment (tmylk) [inline gensim/models/wrappers/varembed.py]: Please subclass KeyedVectors

File: gensim/test/test_varembed_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2017 Anmol Gulati <anmol01gulati@gmail.com>
+# Copyright (C) 2017 Radim Rehurek <radimrehurek@seznam.cz>
+#
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+"""
+Automated tests for checking transformation algorithms (the models package).
```
Comment (tmylk) [inline gensim/test/test_varembed_wrapper.py]: Please change to VarEmbed

File: gensim/test/test_varembed_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    def model_sanity(self, model):
+          # Check vocabulary and vector size
+        self.assertEqual(model.wv.syn0.shape, (model.vocab_size, model.vector_size))
+        self.assertTrue(model.wv.syn0.shape[0] == len(model.wv.vocab))
+
+    def testEnsembleMorphemeEmbeddings(self):
+        """Test ensembling of Morhpeme Embeddings"""
+        model = varembed.VarEmbed.load_varembed_format(vectors=varembed_model_vector_file,
+                                                       morfessor_model=varembed_model_morfessor_file, use_morphemes=True)
+        self.model_sanity(model)
```
Comment (tmylk) [inline gensim/test/test_varembed_wrapper.py]: please test that syn0 is different compared to non-morpheme model.

File: gensim/models/wrappers/varembed.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        D = utils.unpickle(vectors)
+        word_to_ix = D['word_to_ix']
+        morpho_to_ix = D['morpho_to_ix']
+        word_embeddings = D['word_embeddings']
+        morpho_embeddings = D['morpheme_embeddings']
+        result.load_word_embeddings(word_embeddings, word_to_ix)
+        if use_morphemes:
+            try:
+                import morfessor
+                morfessor_model = morfessor.MorfessorIO().read_binary_model_file(morfessor_model)
```
Comment (tmylk) [inline gensim/models/wrappers/varembed.py]: could you raise an issue in varembed github as heads up that `read_binary_model_file` will be [deprecated](https://github.com/aalto-speech/morfessor/issues/5#issuecomment-275996118) by morfessor?

File: gensim/models/wrappers/varembed.py (hunk 2 above)
Comment (anmol01gulati) [inline gensim/models/wrappers/varembed.py]: Done.

File: gensim/models/wrappers/varembed.py (hunk 1 above)
Comment (anmol01gulati) [inline gensim/models/wrappers/varembed.py]: Done.

File: gensim/test/test_varembed_wrapper.py (hunk 3 above)
Comment (anmol01gulati) [inline gensim/test/test_varembed_wrapper.py]: Oh! Had missed this. Thanks. Done.

File: gensim/test/test_varembed_wrapper.py (hunk 4 above)
Comment (anmol01gulati) [inline gensim/test/test_varembed_wrapper.py]: Yes, added now.

File: gensim/models/wrappers/varembed.py (hunk 5 above)
Comment (anmol01gulati) [inline gensim/models/wrappers/varembed.py]: Yes, Sounds good. I've put up an issue  https://github.com/rguthrie3/MorphologicalPriorsForWordEmbeddings/issues/3 to notify them about the new release on morfessor as well.

File: gensim/models/wrappers/varembed.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        if vectors is None:
+            raise Exception(
+                "Please provide vectors binary to load varembed model")
+        D = utils.unpickle(vectors)
+        word_to_ix = D['word_to_ix']
+        morpho_to_ix = D['morpho_to_ix']
+        word_embeddings = D['word_embeddings']
+        morpho_embeddings = D['morpheme_embeddings']
+        result.load_word_embeddings(word_embeddings, word_to_ix)
+        if use_morphemes:
```
Comment (tmylk) [inline gensim/models/wrappers/varembed.py]: just `if morfessor_model` is enough

File: gensim/models/wrappers/varembed.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        logger.info("Corpus has %i words", len(self.vocab))
+        for word_id, word in enumerate(counts):
+            self.vocab[word] = Vocab(index=word_id, count=counts[word])
+            self.syn0[word_id] = word_embeddings[word_to_ix[word]]
+            self.index2word[word_id] = word
+        assert((len(self.vocab), self.vector_size) == self.syn0.shape)
+        logger.info("Loaded matrix of %d size and %d dimensions", self.vocab_size, self.vector_size)
+
+
+    def ensemble_morpheme_embeddings(self, morfessor_model, morpho_embeddings, morpho_to_ix):
```
Comment (tmylk) [inline gensim/models/wrappers/varembed.py]: maybe  `add_morphemes_to_word_embeddings`?

File: gensim/models/wrappers/varembed.py (hunk 6 above)
Comment (anmol01gulati) [inline gensim/models/wrappers/varembed.py]: Yes, sounds good. Done.

File: gensim/models/wrappers/varembed.py (hunk 7 above)
Comment (anmol01gulati) [inline gensim/models/wrappers/varembed.py]: Changed method name now.

**Comment by tmylk** [comment]
Will there be a tutorial ipynb?

**Comment by anmol01gulati** [comment]
Yes, I'll just add one.

**Comment by anmol01gulati** [comment]
@tmylk I've now added a tutorial on varembed model as well. Please review everything once, and I feel we could merge it then.

File: gensim/test/test_varembed_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        self.assertEqual(model.syn0.shape, (model.vocab_size, model.vector_size))
+        self.assertTrue(model.syn0.shape[0] == len(model.vocab))
+
+    @unittest.skipIf(sys.version_info < (2, 7), 'Supported only on Python 2.7 and above')
+    def testAddMorphemesToEmbeddings(self):
+        """Test add morphemes to Embeddings
+           Test only in Python 2.7 and above. Add Morphemes is not supported in earlier versions.
+        """
+        model = varembed.VarEmbed.load_varembed_format(vectors=varembed_model_vector_file)
+        model_with_morphemes = varembed.VarEmbed.load_varembed_format(vectors=varembed_model_vector_file,
```
Comment (piskvorky) [inline gensim/test/test_varembed_wrapper.py]: Code style: hanging indent please (not vertical indent).

File: gensim/test/test_varembed_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                                                                      morfessor_model=varembed_model_morfessor_file)
+        self.model_sanity(model_with_morphemes)
+        # Check syn0 is different for both models.
+        self.assertFalse(np.allclose(model.syn0, model_with_morphemes.syn0))
+
+    @unittest.skipUnless(sys.version_info < (2, 7), 'Test to check throwing exception in Python 2.6 and earlier')
+    def testAddMorphemesThrowsExceptionInPython26(self):
+        self.assertRaises(Exception, varembed.VarEmbed.load_varembed_format, vectors=varembed_model_vector_file,
```
Comment (piskvorky) [inline gensim/test/test_varembed_wrapper.py]: Hanging indent.

File: gensim/test/test_varembed_wrapper.py (hunk 8 above)
Comment (anmol01gulati) [inline gensim/test/test_varembed_wrapper.py]: Done.

File: gensim/test/test_varembed_wrapper.py (hunk 9 above)
Comment (anmol01gulati) [inline gensim/test/test_varembed_wrapper.py]: Changed now.

**Comment by tmylk** [comment]
Thanks for adding the wrapper! It will be part of this week's release.

Let's add a benchmark notebook and a blog in another PR. Will publicize when that is ready.

**Comment by anmol01gulati** [comment]
Cool! Great! That's Awesome! :D 
Yes sure, will open up a new PR for the benchmark and blog documentation.

**Pull request merged**